Republish - Machinery processing refactor

SteelSlayer opened the pull request:
<!-- Write **BELOW** The Headers and **ABOVE** The comments else it may not be viewable. -->
<!-- You can view Contributing.MD for a detailed description of the pull request process. -->

## About The Pull Request
Republish of  #49170

I haven't changed how the processing stuff works that much compared to the old PR, but I learned some new things and I just wanted to start from a clean slate. Still, a majority of the information in that old PR description is still valid.

Just as TL;DR as to what this PR does:
It takes fast processing machines which constantly `process()` and makes it so that they only start processing when they need to, like when they're switched on for example. Some machines now don't even use processing at all, like the ORM. It just listens for ores that enter it's input turf and sucks them up.

## Why It's Good For The Game
This will save a small amount of server resources, because the game won't have to pointlessly execute code when it doesn't need to.

## Changelog
:cl:
code: Updates various machines, which currently use SSfastprocess, so that they only start processing when it's necessary. Some of these updated machines have been changed so that they don't ever need to process.
refactor: Replaces the speed_process var with two new variables. The processing_flags variable stores bitflags with information about a machine's preferences on when it should start processing. The subsystem_type var holds the path to which type of subsystem that machine should use.
/:cl:

<!-- Both :cl:'s are required for the changelog to work! You can put your name to the right of the first :cl: if you want to overwrite your GitHub username as author ingame. -->
<!-- You can use multiple of the same prefix (they're only used for the icon ingame) and delete the unneeded ones. Despite some of the tags, changelogs should generally represent how a player might be affected by the changes rather than a summary of the PR's contents. -->

SteelSlayer committed:
initial

small thing

Qustinnus commented on code/game/machinery/_machinery.dm:
```diff
@@ -137,27 +138,31 @@ Class Procs:
 
 	return INITIALIZE_HINT_LATELOAD
 
+// helper procs for starting and stoppping the appropriate subsystem that a machine uses. Uses their `subsystem_type` var to choose
```
Use 3 / so this is autodocced :)

Qustinnus commented on code/modules/mining/machine_processing.dm:
```diff
@@ -5,6 +5,33 @@
 	var/input_dir = NORTH
 	var/output_dir = SOUTH
+	var/turf/input_turf = null // The turf that the machines will be listening to for ores/items
+	var/needs_item_input = FALSE // whether or not the machine needs to pick up items, i.e. the ORM picking up ores
```
I still think this would be better off in a component that takes a callback.

SteelSlayer replied:
Would this be included in the same component that would handle the I/O?

Qustinnus commented on code/modules/mining/machine_processing.dm:
```diff
@@ -5,6 +5,33 @@
+	input_turf = get_step(src, input_dir) // get the turf that players need to place the ore/items onto, defaults to NORTH at round start
+	if(input_turf && istype(input_turf)) // make sure there is actually a turf
+		// listens for when an atom ENTERS the input_turf, tells the machine to deal with the ores/items
+		RegisterSignal(input_turf, COMSIG_ATOM_ENTERED, .proc/pickup_items, TRUE)
```
Pickup_item it's singular

Qustinnus commented on code/modules/mining/machine_processing.dm:
```diff
@@ -5,6 +5,33 @@
+	if(input_turf)
+		UnregisterSignal(input_turf, COMSIG_ATOM_ENTERED)
+
+/obj/machinery/mineral/proc/pickup_items()
```
Ditto, also you havnt added the parameters which a signal proc should have. Look at another proc linked to this signal for examples

SteelSlayer replied:
Ah yeah, I know the params you're speaking of. I'm curious why you actually need them if the proc gets executed just fine without them though.

Qustinnus replied:
clarity, and in my opinion, you should be using it, it makes it far more event-based rather than looping through the input turf.

SteelSlayer replied:
yeah nevermind I get what you mean now, if Im not looping, I need the signal params, in particular the atom/movable :+1:

Qustinnus commented on code/modules/mining/machine_redemption.dm:
```diff
@@ -149,10 +151,11 @@
-	var/atom/input = get_step(src, input_dir)
+
+	var/atom/input = input_turf
 	var/obj/structure/ore_box/OB = locate() in input
```
Not a fan of this, you're looping through the entire Turf when you can just get the ore that was dropped from this procs params

SteelSlayer replied:
yeah I will make all pickup_item procs across all these machines not loop through the turf and instead just suck up the 1 item that entered.

Qustinnus replied:
👌 awesome

SteelSlayer commented:
@Qustinnus The ORM does not use `process()` anymore. And about the component thing, are you saying the input turf and signal code should be a component and not just on mineral machines?

Qustinnus commented:
My bad the explanation might've not been great, I'm saying you kept the proc intact with it's pre-fix behavior, the signal for atom_enter or whatever it is gives you the AM that entered the turf, this means you can just use that ore instead of looping through the turf when the proc is called.

to your other point; yes, I/O for machinery is not really something I think is exclusive to mining machines but it's not a necesity for this PR, just a reccomendation for code improvement

SteelSlayer committed:
improvements

1. Removes loops for picking up items. its now an item by item basis unless there is an ore box involved
2. Removed pickup_rate and ore_buffer var for the ORM, they aren't needed anymore
3. Fixed conveyors not moving items that get created on top of them (New / Initialize), by sending signals when atoms are created.

SteelSlayer commented:
I don't understand why is travis is failing here. It's complaining about something that compiles fine and works perfectly in game with no issues. I'm just going to leave this as is until someone can tell me whats going on here or I figure it out on my own.

spookydonut commented on code/modules/recycling/conveyor2.dm:
```diff
@@ -55,6 +56,7 @@ GLOBAL_LIST_EMPTY(conveyors_by_id)
 	if(newid)
 		id = newid
 	update_move_direction()
+	RegisterSignal(src, COMSIG_MOVABLE_CROSSED, .Crossed) // For receiving signals from newly created items. These do not call Crossed() so it needs a signal
```
```suggestion
	RegisterSignal(src, COMSIG_MOVABLE_CROSSED, .proc/Crossed) // For receiving signals from newly created items. These do not call Crossed() so it needs a signal
```

SteelSlayer replied:
Yeah it is cursed. I'm just changing the proc name to `conveyorCrossed`, which works fine

spookydonut reviewed (changes requested):
`.procname` is C U R S E D

always use at least `.proc/procname` to avoid C U R S E D behaviour

SteelSlayer committed:
renames the registered signal proc name so travis doesn't throw a fit

ninjanomnom commented on code/game/atoms.dm:
```diff
@@ -145,6 +145,11 @@
+	if(loc)
+		SEND_SIGNAL(loc, COMSIG_ATOM_ENTERED, src) // if the new item is created over a turf, send a signal that it entered that turf, `loc` in this case
+		for(var/atom/A in loc) // send a signal for every item on the `loc` turf to let them know a new item just crossed it
+			SEND_SIGNAL(A, COMSIG_MOVABLE_CROSSED, src)
```
Just send on the turf, things in the turf can register on the turf if they need to know this

SteelSlayer replied:
I'm not sure I understand what you mean here.

ninjanomnom replied:
You don't need to send the signal to all the contents because anything in those contents can listen to the signal on their loc instead

SteelSlayer replied:
Ok so if I understand, I should ditch these two lines here, but keep the one above it for ATOM_ENTERED. Then in my case I would just register the conveyor to listen for that signal on its `loc`?

ninjanomnom replied:
yes

SteelSlayer committed:
signal improvement

SteelSlayer committed:
forgot to update other proc names

ninjanomnom commented on code/__DEFINES/machines.dm:
```diff
@@ -12,6 +12,10 @@
+// bitflags for a machine's preferences on when it should start processing
+#define START_PROCESSING_ON_INIT	(1<<0)	// if the machine should start processing on Initialize()
+#define START_PROCESSING_MANUALLY	(1<<1)	// if the machine will start processing in the future, after Initialize(), such as when a certain condition is met
+#define NEVER_PROCESS				(1<<2)	// if the machine has no reason to process
```
What's the difference between manual and never, there doesn't seem to be a reason to have both

SteelSlayer replied:
You're right there's no difference, I created it simply to make it clear that the machine is not and should never process. I can drop the never and just use manual instead if you want.

ninjanomnom replied:
Yeah it's not really useful to have both

ninjanomnom commented on code/modules/mining/machine_redemption.dm:
```diff
@@ -149,29 +140,29 @@
+
+	if(!console_notify_timer)
+		// gives 5 seconds for a load of ores to be sucked up by the ORM before it sends out request console notifications. This should be enough time for most deposits that people make
+		console_notify_timer = addtimer(CALLBACK(src, .proc/send_console_message), 50)
```
```suggestion
		console_notify_timer = addtimer(CALLBACK(src, .proc/send_console_message), 5 SECONDS)
```

ninjanomnom commented on code/modules/mining/machine_redemption.dm:
```diff
@@ -149,29 +140,29 @@
 
 /obj/machinery/mineral/ore_redemption/attackby(obj/item/W, mob/user, params)
 	if(default_unfasten_wrench(user, W))
+		if(anchored)
```
override default_unfasten_wrench

SteelSlayer committed:
ninjanomnom review changes

replace NEVER_PROCESS with START_PROCESSING_MANUALLY

default_unfasten_wrench override for ORM

50 -> 5 SECONDS

I totally didn't mispell anything

ninjanomnom commented on code/game/atoms.dm:
```diff
@@ -145,6 +145,9 @@
 	flags_1 |= INITIALIZED_1
 
+	if(loc)
+		SEND_SIGNAL(loc, COMSIG_ATOM_ENTERED, src) // if the new item is created over a turf, send a signal that it entered that turf, `loc` in this case
```
I believe I said before you need to use a new signal, this is not the entered proc

SteelSlayer replied:
Oh do mean I should make a new signal name, `COMSIG_ATOM_CREATED` or something like that?

ninjanomnom replied:
Sure

SteelSlayer commented:
I think I did an oopsie when updating my branch from master. I not exactly sure what just happened, but it triggered gitdiff to pick up the entire ` code/__DEFINES/dcs/signals.dm` file, when I only added one new signal on line 50: `#define COMSIG_ATOM_CREATED "atom_created"`. I'm not really sure how to fix this from here.

SteelSlayer committed:
makes a new signal: COMSIG_ATOM_CREATED

SteelSlayer committed:
more review changes

ninjanomnom commented on code/modules/mining/machine_processing.dm:
```diff
@@ -5,6 +5,34 @@
+
+/obj/machinery/mineral/proc/unregister_input_turf()
+	if(input_turf)
+		UnregisterSignal(input_turf, COMSIG_ATOM_ENTERED)
```
```suggestion
		UnregisterSignal(input_turf, list(COMSIG_ATOM_ENTERED, COMSIG_ATOM_CREATED))
```

SteelSlayer replied:
ugh, how could I forget..

SteelSlayer committed:
duh

Rohesie commented on code/modules/mining/machine_processing.dm:
```diff
@@ -5,6 +5,34 @@
+	if(input_turf) // make sure there is actually a turf
+		// listens for when an atom ENTERS the input_turf, tells the machine to deal with the ores/items
+		RegisterSignal(input_turf, COMSIG_ATOM_ENTERED, .proc/pickup_item)
+		RegisterSignal(input_turf, COMSIG_ATOM_CREATED, .proc/pickup_item)
```
```suggestion
		RegisterSignal(input_turf, list(COMSIG_ATOM_ENTERED, COMSIG_ATOM_CREATED), .proc/pickup_item)
```

Rohesie commented on code/modules/mining/machine_processing.dm:
```diff
@@ -142,8 +167,15 @@
+		return
+
+	var/obj/item/stack/ore/O = target
+	process_ore(O)
```
No need to type cast this here. You could simply do `process_ore(target)`

Rohesie commented on code/modules/recycling/conveyor2.dm:
```diff
@@ -55,6 +56,8 @@ GLOBAL_LIST_EMPTY(conveyors_by_id)
 		id = newid
 	update_move_direction()
+	RegisterSignal(loc, COMSIG_ATOM_CREATED, .proc/conveyorCrossed) // For receiving signals from newly created items.
+	RegisterSignal(loc, COMSIG_ATOM_ENTERED, .proc/conveyorCrossed) // When any atom moves onto the location of this belt
```
```suggestion
	RegisterSignal(loc, list(COMSIG_ATOM_CREATED, COMSIG_ATOM_ENTERED), .proc/conveyorCrossed) // For receiving signals from newly created items.
```

Rohesie commented on code/modules/recycling/conveyor2.dm:
```diff
@@ -123,35 +126,43 @@ GLOBAL_LIST_EMPTY(conveyors_by_id)
-	//get the first 30 items in contents
+// returns true if there is at least 1 item on the belt that can be moved
+/obj/machinery/conveyor/proc/check_belt_contents()
+	for(var/atom/movable/AM in loc.contents)
```
Just to note, `contents` is implied.
This works the same as `for(var/atom/movable/AM in loc)`

Rohesie commented on code/modules/recycling/conveyor2.dm:
```diff
@@ -123,35 +126,43 @@ GLOBAL_LIST_EMPTY(conveyors_by_id)
+// returns true if there is at least 1 item on the belt that can be moved
+/obj/machinery/conveyor/proc/check_belt_contents()
+	for(var/atom/movable/AM in loc.contents)
+		if(AM && !AM.anchored)
```
```suggestion
		if(!AM.anchored)
```
If it gets found by the `for` then it must exist. No need to check for it.

SteelSlayer replied:
Yeah... I'm not sure why I did this. Thanks for catching all this silly stuff.

Rohesie commented on code/modules/recycling/conveyor2.dm:
```diff
@@ -123,35 +126,43 @@ GLOBAL_LIST_EMPTY(conveyors_by_id)
-	conveying = TRUE
-	addtimer(CALLBACK(src, .proc/convey, affecting), 1)
+	for(var/atom/movable/AM in loc.contents)
+		if(AM && !AM.anchored)
```
```suggestion
		if(!AM.anchored)
```

Rohesie commented on code/modules/recycling/conveyor2.dm:
```diff
@@ -123,35 +126,43 @@ GLOBAL_LIST_EMPTY(conveyors_by_id)
-	addtimer(CALLBACK(src, .proc/convey, affecting), 1)
+	for(var/atom/movable/AM in loc.contents)
+		if(AM && !AM.anchored)
+			i += 1
```
```suggestion
			i++
```

Rohesie commented on code/modules/recycling/conveyor2.dm:
```diff
@@ -123,35 +126,43 @@ GLOBAL_LIST_EMPTY(conveyors_by_id)
+	for(var/atom/movable/AM in loc.contents)
+		if(AM && !AM.anchored)
+			i += 1
+			if(i >= MAX_CONVEYOR_ITEMS_MOVE)
```
or this:
```suggestion
			if(++i >= MAX_CONVEYOR_ITEMS_MOVE)
```
(don't do this if you do the previous suggestion)

SteelSlayer committed:
even more review improvements

Rohesie commented on code/modules/recycling/conveyor2.dm:
```diff
@@ -123,35 +125,43 @@ GLOBAL_LIST_EMPTY(conveyors_by_id)
 			continue
-		i++ // we're sure it's a real target to move at this point
-		if(i >= MAX_CONVEYOR_ITEMS_MOVE)
+		if(++i >= MAX_CONVEYOR_ITEMS_MOVE)
```
In order to do what you are suggesting to do you'll need either of these:
`if(++i > MAX_CONVEYOR_ITEMS_MOVE)`
or
`if(i++>= MAX_CONVEYOR_ITEMS_MOVE)`
Else you'll be moving one less than the `MAX_CONVEYOR_ITEMS_MOVE`
If, say, the maximum is `1`, it will move none.
That is the old logic, not introduced by your PR from what I see.

SteelSlayer replied:
oh yeah, you're right, gonna go with `++i >`

Rohesie replied:
That works. Thanks for the fix!

SteelSlayer committed:
switch >= to >

Rohesie commented on code/modules/mining/machine_processing.dm:
```diff
@@ -5,6 +5,33 @@
+/obj/machinery/mineral/Destroy()
+	. = ..()
+	if(input_turf)
+		unregister_input_turf()
```
On deletion things will unregister themselves clearly already.
You'll only need to care about things registered to the thing being deleted.

SteelSlayer replied:
I'm confused by what you mean here, can you re-phrase?

Qustinnus replied:
destroy() will automatically handle you unregister to any events, but there might still be things listening to events of the qdel'd object

SteelSlayer replied:
Ah, so basically I don't even need this Destroy override, the signal should unregister itself when the machine is destroyed?

Rohesie replied:
Yep. When something is destroyed it will automatically unregister from everything.
And actually, to correct myself, it will also unregister properly things listening to it, so no need to worry about it.
You can remove this from `Destroy()` safely.

Rohesie commented on code/modules/recycling/conveyor2.dm:
```diff
@@ -55,6 +56,7 @@ GLOBAL_LIST_EMPTY(conveyors_by_id)
 	if(newid)
 		id = newid
 	update_move_direction()
+	RegisterSignal(loc, list(COMSIG_ATOM_ENTERED, COMSIG_ATOM_CREATED), .proc/conveyorCrossed) // When any atom moves onto the belt or gets created over the belt.
```
Think you could add an `if(operating)` check, and register when turning on and unregister when turning off?

SteelSlayer replied:
Yeah I can do that, will be more efficient probably.

Rohesie commented on code/modules/recycling/conveyor2.dm:
```diff
@@ -277,24 +286,28 @@ GLOBAL_LIST_EMPTY(conveyors_by_id)
 		C.operating = position
 		C.update_move_direction()
 		C.update_icon()
+		if(abs(C.operating)) // equals 1 if the conveyor is moving in either direction, 0 if turned off
```
Both `-1` and `1` are truthy values. `abs()` makes sense but it's not really needed.

Rohesie commented on code/modules/recycling/conveyor2.dm:
```diff
@@ -123,35 +125,43 @@ GLOBAL_LIST_EMPTY(conveyors_by_id)
+	if(!operating || machine_stat & (BROKEN | NOPOWER))
+		return
+
+	get_belt_contents() // populate affecting with the movable items on the belt
```
Should this be called every time a movable atom enters said turf?
It looks like this creates a lot of redundancy.
You might want to handle each entered atom by itself instead of triggering a `loc` contents check every time.

SteelSlayer replied:
It is probably better to do it on an atom by atom basis yeah. I think I kept this around because there's suppose to be a 30 item move limit and I didn't want to mess with that, but I can take another look.

Rohesie replied:
The problem here is that this will be triggered every time an item moves here, so it's potentially happeing a lot of times per second without checks, unlike before.

SteelSlayer replied:
Now I remember why I did it this way. This is to solve an issue where, if you for some reason stop over a conveyor belt that's running, say you run into some plastic flaps for example. And then you rest to continue getting pushed, the belt wouldn't push you, because the belt pushing is event based, only getting triggered when something moves over it.

The way I have it now, while not ideal, ensures that things sitting on belts and not moving, get moving when they are able. I'm not really sure what to do about this issue.

Rohesie replied:
The solution is to account for the event of the density (or anchored, or other conditions that prevent moving) changing.

SteelSlayer replied:
Well I messed around with adding more handling for these events, and had it all working fine. However when I profiled it compared to just using the standard `process()` instead, signal based conveying is actually more expensive.

I think I'll just revert back to use `process()` for conveyors, but still keep some of the code improvements I've made.

Rohesie replied:
It's probably more expensive but also more responsive, yes.

SteelSlayer committed:
reverts conveyors back to using process() for moving stuff

Dennok commented on code/modules/recycling/conveyor2.dm:
```diff
@@ -125,43 +124,39 @@ GLOBAL_LIST_EMPTY(conveyors_by_id)
-		if(AM.anchored) // If it can't be moved, ignore it
+	for(var/item in loc.contents)
+		if(item == src)
 			continue
```
Try this
```suggestion
	for(var/item in loc.contents-src)
```
Drip of unload

SteelSlayer replied:
Yeah you're right I should fix that while I'm here. Doing `if(item == src)` is gross.

Rohesie commented on code/modules/mining/machine_processing.dm:
```diff
@@ -5,6 +5,28 @@
+	var/turf/input_turf = null // The turf that the machines will be listening to for ores/items
+	var/needs_item_input = FALSE // whether or not the machine needs to pick up items, i.e. the ORM picking up ores
+	processing_flags = START_PROCESSING_MANUALLY
+	subsystem_type = /datum/controller/subsystem/processing/fastprocess
```
Put the redefined variables at the t op of the object definition, before the object variables, please.

SteelSlayer replied:
Sure. also yeah I'll write some documentation

Dennok commented on code/modules/recycling/conveyor2.dm:
```diff
@@ -125,43 +124,39 @@ GLOBAL_LIST_EMPTY(conveyors_by_id)
-	get_belt_contents() // populate affecting with the movable items on the belt
-	addtimer(CALLBACK(src, .proc/convey, affecting), pick(1,2)) // pick(1,2) so players have a bit of trouble walking over belts moving in the opposite direction they are
+	conveying = TRUE
+	addtimer(CALLBACK(src, .proc/convey, affecting), 1)
```
I think this must be, and before with signal trigger, unique, so no new thigger don't create new timer when old dont fire.
Or owerride

I think it can be source of increased load

SteelSlayer replied:
This timer will never fail to fire. `process()` is ran 5 times a second (from SSfastprocess) and nothing can increase this frequency. If `process()` returns early, the timer and callback are never created. Otherwise the timer/callback will always fire and therefore it will be unique.

Dennok commented on code/modules/recycling/conveyor2.dm:
```diff
@@ -125,43 +124,39 @@ GLOBAL_LIST_EMPTY(conveyors_by_id)
 		if(!QDELETED(A) && (A.loc == loc))
 			A.ConveyorMove(movedir)
 			//Give this a chance to yield if the server is busy
 			stoplag()
```
I think `if(!QDELETED(A) && (A.loc == loc))` can be removed if use `for(var/atom/movable/A in affecting&loc.contents)`
QDELETED items must be in nullspace so it no in contents of our turf.

```suggestion
	for(var/atom/movable/A in affecting&loc.contents)
		A.ConveyorMove(movedir)
		//Give this a chance to yield if the server is busy
		stoplag()
```

As i think about this. 
It need check like "its posible to move something?" before we try to move all things, 
or we create load generator,
 by place full conveyor directed to wall.

Dennok replied:
I think `if(!QDELETED(A) && (A.loc == loc))` can be removed if use `for(var/atom/movable/A in affecting&loc.contents)`
QDELETED items must be in nullspace so it no in contents of our turf.

```suggestion
	for(var/atom/movable/A in affecting&loc.contents)
		A.ConveyorMove(movedir)
		//Give this a chance to yield if the server is busy
		stoplag()
```


As i think about this. 
It need check like "its posible to move something?" before we try to move all things, 
or we create load generator,
 by place full conveyor directed to wall.
But while we has limit to 30 things its must be small load.

Dennok commented on code/modules/recycling/conveyor2.dm:
```diff
@@ -125,43 +124,39 @@ GLOBAL_LIST_EMPTY(conveyors_by_id)
-	// if the conveyor isn't switched on or is broken / has no power
-	if(!operating || machine_stat & (BROKEN | NOPOWER))
-		return
+		affecting.Add(item)
```
As i think, and that already written upper

`conveyor Crossed` in trigger system must be used to dirrectly populate contents of `affecting` instead of call get_belt_contents() that parce all contents of our turf EVERY TIME when something step on conveyor.
And it create new timer for EVERY THING that step on conveyor.

OR

Make system for individual conveyor item transferring.
`conveyor Crossed` create timer that try move only triger thing after 1 tick as it cross in if thing still in our turf.

SteelSlayer replied:
I think you may be looking at older code. I recently reverted back to using the old way of moving stuff for belts, which is using `process()`. I'm not doing signal based stuff anymore for the conveyor because to get it working like the old system, it ended up being more expensive than how the belts currently handle it.

Dennok replied:
I read this.
I write what i think can make trigger system less expensive

Dennok commented:
what use less of power?

if(thing in T.contents) // i think it depents of contents size

if(T.contents.Find(thing)) //same?

if(thing.loc = T) //work if it must be dirrectly in turf

if(get_turf(thing) = T) //work if it can be in something that in turf

SteelSlayer committed:
various touch ups, adds documentation

SteelSlayer commented:
For now I'm going to leave the conveyors mostly untouched. I may try to revisit them later and see what I can do with signals but not in this PR. This is merge ready, provided there are no other issues.

SteelSlayer committed:
rebase to fix map conflicts with forgottenship.dmm
